tests/e2e/datadog/logs/mod.rs

```diff
@@ -0,0 +1,72 @@
+use serde_json::Value;
+
+use vector::test_util::trace_init;
+
+use super::*;
+
+const LOGS_ENDPOINT: &str = "/api/v2/logs";
+
+fn expected_log_events() -> usize {
+    std::env::var("EXPECTED_LOG_EVENTS")
+        .map(|n_expected| {
+            n_expected
+                .parse::<usize>()
+                .expect("EXPECTED_LOG_EVENTS should be an unsigned integer.")
+        })
+        .unwrap_or(1000)
+}
+
+// Asserts that each log event has the hostname and timestamp fields, and
+// Removes them from the log so that comparison can more easily be made.
+// @return the number of log entries in the payload.
+fn assert_timestamp_hostname(payloads: &mut [Value]) -> usize {
+    let mut n_log_events = 0;
+
+    payloads.iter_mut().for_each(|payload_array| {
+        payload_array
+            .as_array_mut()
+            .expect("should be array")
+            .iter_mut()
+            .for_each(|log_val| {
+                n_log_events += 1;
+
+                let log = log_val
+                    .as_object_mut()
+                    .expect("log entries should be objects");
+
+                assert!(log.remove("timestamp").is_some());
+                assert!(log.remove("hostname").is_some());
+            })
+    });
+
+    n_log_events
+}
+
+// runs assertions that each set of payloads should be true to regardless
+// of the pipeline
+fn common_assertions(payloads: &mut [Value]) {
+    assert!(payloads.len() > 0);
+
+    let n_log_events = assert_timestamp_hostname(payloads);
+
+    println!("log events received: {n_log_events}");
+
+    assert!(n_log_events == expected_log_events());
+}
+
+#[tokio::test]
+async fn validate() {
+    trace_init();
+
+    println!("getting log payloads from agent-only pipeline");
+    let mut agent_payloads = get_payloads_agent(LOGS_ENDPOINT).await;
+
+    common_assertions(&mut agent_payloads);
+
+    println!("getting log payloads from agent-vector pipeline");
+    let mut vector_payloads = get_payloads_vector(LOGS_ENDPOINT).await;
+
+    common_assertions(&mut vector_payloads);
+
+    assert_eq!(agent_payloads, vector_payloads);
```

Comment:
Will this comparison always succeed? What if Vector batches logs into requests differently than how they were received from the agent? Will requests ever be received out of order? If yes to any of those, we may want to normalize the payloads here to prevent flakiness.

Reply:
Definitely a valid concern. I chose settings in the Agent and Vector configs , as well as the log generation to attempt to result in this not happening. I just ran it in a loop for 15 minutes and it never failed.

Another thing to consider- I think this type of scenario will be more likely to be a concern for the metrics and or traces e2e tests. So I'm already anticipating potentially needing to do some of that there.

If you have strong opinions about it I can add it pre emptively here, but given my local testing I think the risk is low for flakiness from the logs e2e tests.